channel/ledger/channel.go

```diff
@@ -0,0 +1,244 @@
+		keys = append(keys, k)
+	}
+	sort.Slice(keys, func(i, j int) bool {
+		return keys[i].String() < keys[j].String()
+	})
+
+	for _, target := range keys {
+		allocations = append(allocations, o.guarantees[target].AsAllocation())
+
+	}
+
+	return outcome.Exit{
+		outcome.SingleAssetExit{
+			Asset:       o.assetAddress,
+			Allocations: allocations,
+		},
+	}
+}
+
+var ErrInsufficientFunds = fmt.Errorf("unable to divert to guarantee: insufficient funds")
+var ErrDuplicateGuarantee = fmt.Errorf("duplicate guarantee detected")
+
+// DivertToGuarantee deducts g.amount from o.left's balance, and
+// adds g to o.guarantees
+func (o LedgerOutcome) DivertToGuarantee(g Guarantee) (LedgerOutcome, error) {
+
+	if g.amount.Cmp(&o.left.amount) == 1 {
+		return LedgerOutcome{}, ErrInsufficientFunds
+	}
+
+	o.left.amount.Sub(&o.left.amount, &g.amount)
+
+	_, found := o.guarantees[g.target]
+	if found {
+		return LedgerOutcome{}, ErrDuplicateGuarantee
+	}
+	o.guarantees[g.target] = g
+
+	return o, nil
+}
+
+// Vars stores the turn number and outcome for a state in a consensus channel
+type Vars struct {
+	TurnNum uint64
+	Outcome LedgerOutcome
+}
+
+// SignedVars stores 0-2 signatures for some vars in a consensus channel
+type SignedVars struct {
+	Vars
+	Signatures [2]*state.Signature
+}
+
+// SignedProposal is a proposal with a signature on it
+type SignedProposal struct {
+	state.Signature
+	Proposal interface{}
+}
+
+// Add is a proposal to add a guarantee for the given virtual channel
+// amount is to be deducted from left
+type Add struct {
+	turnNum uint64
+	Guarantee
+}
+
+// Remove is a proposal to remove a guarantee from the ledger channel
+// - `paid` funds are added to the right participant's balance
+// - the remainder are returned to the left participant's balance
+type Remove struct {
+	TurnNum uint64
+	VId     types.Destination
+	Paid    types.Funds
+}
+
+var ErrIncorrectTurnNum = fmt.Errorf("incorrect turn number")
+
+// Add updates Vars by including a guarantee, updating balances accordingly
+func (vars Vars) Add(p Add) (Vars, error) {
+	if p.turnNum != vars.TurnNum+1 {
+		return Vars{}, ErrIncorrectTurnNum
+	}
+
+	vars.TurnNum += 1
+
+	o, err := vars.Outcome.DivertToGuarantee(p.Guarantee)
+
+	if err != nil {
+		return Vars{}, err
+	}
+
+	vars.Outcome = o
+	return vars, nil
+}
+
+func (vars Vars) Remove(p Remove) (Vars, error) {
+	panic("UNIMPLEMENTED")
+}
+
+// Add receives a Guarantee, and generates and stores a SignedProposal in
+// the queue, returning the resulting SignedProposal
+func (c *ConsensusChannel) Add(g Guarantee, sk []byte) (SignedProposal, error) {
+	if c.MyIndex != proposerIndex {
+		return SignedProposal{}, fmt.Errorf("only proposer can call Add")
+	}
+
+	vars := c.current.Vars
+	var err error
+
+	for _, p := range c.proposalQue {
+		vars, err = vars.Add(p.Proposal.(Add))
+		if err != nil {
+			return SignedProposal{}, err
+		}
+	}
+
+	latest := c.proposalQue[len(c.proposalQue)-1].Proposal.(Add)
```

Comment:
This type assertion is a potential footgun. You might want to add some error handling (you can extract an `ok` bool as a second argument after `latest`), or a TODO at least.

Reply:
What is a the footgun in this case?

Reply:
The problem is that you are telling the compiler that you _know better_ about the type of the proposal. If it is not in fact an `Add` -- which seems very likely to happen once this code evolves a bit -- you are likely to cause a panic at runtime. 

If you do 

```go
latest, ok := c.proposalQue[len(c.proposalQue)-1].Proposal.(Add)
```

you can inspect `ok` to see if the type assertion succeeded, and handle that situation gracefully. Another option would be to use a [type switch](https://go.dev/tour/methods/16).

Reply:
I'd expect to use a type switch in the future, once we implement virtual-defunding and ledger top-ups!

Reply:
34f736c902ad51c37ce8e0af267ca931543d4b9b